src/Language/Mimsa/Typechecker/Codegen/Functor.hs

```diff
@@ -6,84 +6,141 @@ module Language.Mimsa.Typechecker.Codegen.Functor
   )
 where
 
+import Control.Monad.Except
+import Control.Monad.State
 import Data.Foldable (foldl')
 import qualified Data.List.NonEmpty as NE
 import Data.Map (Map)
 import qualified Data.Map as M
+import Data.Semigroup
 import Data.Text (Text)
+import Language.Mimsa.Printer
 import Language.Mimsa.Types.AST
 import Language.Mimsa.Types.Identifiers
 import Prelude hiding (fmap)
 
+type FunctorM = StateT (Map Name Int) (Either Text)
+
+runFunctorM :: FunctorM a -> Either Text a
+runFunctorM fn = case runStateT fn mempty of
+  Right (a, _) -> pure a
+  Left e -> Left e
+
+functorMap :: DataType -> Either Text (Expr Name ())
+functorMap = runFunctorM . functorMap_
+
 -- | A newtype is a datatype with one constructor
 -- | with one argument
-functorMap ::
+functorMap_ ::
   DataType ->
-  Either Text (Expr Name ())
-functorMap (DataType tyCon vars items) = do
+  FunctorM (Expr Name ())
+functorMap_ (DataType tyCon vars items) = do
   let tyName = tyConToName tyCon
   fVar <- getFunctorVar vars
   case getMapItems items of
-    Nothing -> Left "Type should have at least one constructor"
+    Nothing -> throwError "Type should have at least one constructor"
     Just constructors -> do
       matches <-
         traverse
           ( \(consName, fields) ->
-              (,) consName <$> createMatch fVar consName fields
+              (,) consName <$> createMatch tyCon fVar consName fields
           )
           constructors
       pure
-        ( MyLambda
+        ( MyLet
             mempty
-            "f"
+            "fmap"
             ( MyLambda
                 mempty
-                tyName
-                ( MyCaseMatch
+                "f"
+                ( MyLambda
                     mempty
-                    (MyVar mempty tyName)
-                    matches
-                    Nothing
+                    tyName
+                    ( MyCaseMatch
+                        mempty
+                        (MyVar mempty tyName)
+                        matches
+                        Nothing
+                    )
                 )
             )
+            (MyVar mempty "fmap")
         )
 
-getFunctorVar :: [Name] -> Either Text Name
+getFunctorVar :: [Name] -> FunctorM Name
 getFunctorVar names = case NE.nonEmpty names of
-  Just neNames -> Right $ NE.last neNames
-  _ -> Left "Type should have at least one type variable"
+  Just neNames -> pure $ NE.last neNames
+  _ -> throwError "Type should have at least one type variable"
+
+data FieldItemType
+  = VariableField Name
+  | RecurseField Name
+
+toFieldItemType :: TyCon -> Field -> FunctorM FieldItemType
+toFieldItemType typeName = \case
+  VarName a -> pure (VariableField a)
+  ConsName fieldConsName _fields
```

Comment:
Currently we can only recurse through the current datatype. Next I suppose will be mixing them up so we can express Rose Trees etc.

```data Tree a = Tree a [Tree a]```

Which becomes

`type RoseTree a = RoseTree a (List (RoseTree a))`

which does typecheck, fortunately.